adds (first) EventProcessor for system info

fetzig opened the pull request:
* created GlobalEventProcessor to retain eventprocessor instances.
 * implemented trigger to execute eventprocessors on events that are processed
 * Moved context creation into a EventProcessor
 * dropped context creation in `SentryCrashReportConverter`

tested with ios testapp on simulator. context/system info of messages and crash events look like before this PR on sentry.io.

Open points:
 *  What is`otherContexts`?

HazAT commented on Sources/Sentry/SentryClient.m:
```diff
@@ -102,6 +109,7 @@ - (BOOL)checkSampleRate:(NSNumber *)sampleRate {
 - (SentryEvent *_Nullable)prepareEvent:(SentryEvent *)event
                              withScope:(SentryScope *)scope {
     NSParameterAssert(event);
+    NSLog(@"prepareEvent");
```
Please remove, we don't want any `NSLog` statements since you can't control them.

fetzig replied:
Done

HazAT commented on Sources/Sentry/SentryClient.m:
```diff
@@ -130,13 +137,13 @@ - (SentryEvent *_Nullable)prepareEvent:(SentryEvent *)event
-    return event;
-}
 
+    return [scope applyToEvent:event];
```
`applyToEvent:event` must be called before `beforeSend` reason is that the user should get the fully built event before we send it.

fetzig replied:
Moved `beforeSend` call to `SentryClient.captureEvent`.

HazAT commented on Sources/Sentry/SentryCrashIntegration.m:
```diff
@@ -29,10 +37,25 @@ @interface SentryCrashIntegration ()
+}
+
 - (BOOL)installWithOptions:(nonnull SentryOptions *)options {
+    NSLog(@"installWithOptions");
```
remove `NSLog`

fetzig replied:
Done

HazAT commented on Sources/Sentry/SentryCrashIntegration.m:
```diff
@@ -49,30 +72,122 @@ - (BOOL)startCrashHandlerWithError:(NSError *_Nullable *_Nullable)error {
 }
 
+- (void)addEventProcessor {
+    NSLog(@"addEventProcessor");
```
Remove `NSLog`

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryScope.m:
```diff
@@ -134,16 +137,35 @@ - (void)applyToEvent:(SentryEvent *)event {
 }
 
+- (SentryEvent *)callEventProcessors:(SentryEvent *)event {
+    NSLog(@"callEventProcessors");
```
Remove `NSLog`

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryScope.m:
```diff
@@ -99,7 +101,8 @@ - (void)clearBreadcrumbs {
 
-- (void)applyToEvent:(SentryEvent *)event {
+- (SentryEvent * __nullable)applyToEvent:(SentryEvent *)event {
+    NSLog(@"applyToEvent");
```
Remove `NSLog`

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryScope.m:
```diff
@@ -134,16 +137,35 @@ - (void)applyToEvent:(SentryEvent *)event {
+    NSLog(@"SentryGlobalEventProcessor.shared.processors %d", SentryGlobalEventProcessor.shared.processors.count);
+
+    for (SentryEventProcessor processor in SentryGlobalEventProcessor.shared.processors) {
+        NSLog(@"callEventProcessors looop");
```
Remove `NSLog`

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryScope.m:
```diff
@@ -134,16 +137,35 @@ - (void)applyToEvent:(SentryEvent *)event {
+    NSLog(@"callEventProcessors");
+    SentryEvent *newEvent = event;
+
+    NSLog(@"SentryGlobalEventProcessor.shared.processors %d", SentryGlobalEventProcessor.shared.processors.count);
```
Remove `NSLog`, we might want to use `SentryLog` for things that make sense to debug for the user.
In JS also, for example, we tell if an `eventProcessor` returned `null` so we print it we exit the loop early.

fetzig replied:
done

HazAT commented on Sources/Sentry/include/SentryCrashIntegration.h:
```diff
@@ -20,6 +20,13 @@ NS_ASSUME_NONNULL_BEGIN
+ * @return NSDictionary system info.
+ */
++ (NSDictionary *)systemInfo;
+
```
I don't think we need to expose this publicly

fetzig replied:
done

HazAT commented on Sources/Sentry/include/SentryIntegrationProtocol.h:
```diff
@@ -24,11 +24,13 @@ NS_ASSUME_NONNULL_BEGIN
 
+@optional
+- (SentryEvent *)processEvent:(SentryEvent *)event;
+
```
I think this can be removed.

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryCrashIntegration.m:
```diff
@@ -49,30 +72,122 @@ - (BOOL)startCrashHandlerWithError:(NSError *_Nullable *_Nullable)error {
+    SentryEventProcessor eventProcessor = ^SentryEvent * _Nullable(SentryEvent * _Nonnull event) {
+
+        // skip early if integration (and therefore this event processor) isn't active on current client
+        if (NO == [SentrySDK.currentHub isIntegrationActiveInBoundClient:@"SentryCrashIntegration"]) {
```
Can't we use `SentryCrashIntegration.class` here instead of a string?
This is not very safe when it comes to refactoring.

fetzig replied:
forgot that as well. done

HazAT commented on Sources/Sentry/SentryCrashIntegration.m:
```diff
@@ -7,16 +7,24 @@
 
 #if __has_include(<Sentry/Sentry.h>)
-
+#import <UIKit/UIKit.h>
```
This is dangerous, this should be in the `#if __has_include(<Sentry/Sentry.h>)`
also; if you build this on Mac this will fail since it's not there, we need to wrap this in 
```
#if SENTRY_HAS_UIKIT
#import <UIKit/UIKit.h>
#endif
```

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryGlobalEventProcessor.m:
```diff
@@ -0,0 +1,38 @@
+}
+
+- (BOOL)addEventProcessor:(SentryEventProcessor)newProcessor {
+    NSLog(@"addEventProcessor before %lu", (unsigned long)self.processors.count);
```
`NSLog`

fetzig replied:
done

HazAT commented on Sources/Sentry/SentryGlobalEventProcessor.m:
```diff
@@ -0,0 +1,38 @@
+- (BOOL)addEventProcessor:(SentryEventProcessor)newProcessor {
+    NSLog(@"addEventProcessor before %lu", (unsigned long)self.processors.count);
+    [self.processors addObject:newProcessor];
+    NSLog(@"addEventProcessor after %lu", (unsigned long)self.processors.count);
```
`NSLog`

fetzig replied:
done

HazAT commented:
wrt `otherContexts`, something broke here.
The idea is the user can set `setContext` in scope or on the event, this should get merged into `event.contexts` and not `event.context.otherContexts`

So if a user does `setContext(@"abcd", @{@"a": @"b"})`; it should become
`event.contexts.abcd = @{@"a": @"b"}`
and not 
`event.contexts.otherContexts.abcd = @{@"a": @"b"}`

fetzig commented:
> wrt `otherContexts`, something broke here.
> The idea is the user can set `setContext` in scope or on the event, this should get merged into `event.contexts` and not `event.context.otherContexts`
> 
> So if a user does `setContext(@"abcd", @{@"a": @"b"})`; it should become
> `event.contexts.abcd = @{@"a": @"b"}`
> and not
> `event.contexts.otherContexts.abcd = @{@"a": @"b"}`

I suspect you mean `event.context` and not `event.contextS`. correct?

have no idea how I came up with this nonsense-code. fixed and refactored this a little bit and added a comment. renamed it from `otherContexts` to `customContext`. code that was responsible to merge this into `event.context` pevailed. somehow i added this nonsense while implementing scope. updated `Scope.setContextValue..` to set values in `scope.context.cutsomContext` (was `scope.context`).
